pkg/collector/sharedlibrary/library_loader.go

```diff
@@ -0,0 +1,164 @@
+	Version(versionPtr *C.version_function_t) (string, error)
+}
+
+type sharedLibraryLoader struct {
+	folderPath string
+	aggregator *C.aggregator_t
+}
+
+// Load looks for a shared library with the corresponding name and check if it has a `Run` symbol.
+// If that's the case, then the method will return handles for both.
+func (l *sharedLibraryLoader) Load(name string) (libraryHandles, error) {
+	var cErr *C.char
+
+	// the prefix "libdatadog-agent-" is required to avoid possible name conflicts with other shared libraries in the include path
+	libPath := path.Join(l.folderPath, "libdatadog-agent-"+name+getLibExtension())
+
+	cLibPath := C.CString(libPath)
+	defer C.free(unsafe.Pointer(cLibPath))
+
+	cLibHandles := C.load_shared_library(cLibPath, &cErr)
+	if cErr != nil {
+		defer C.free(unsafe.Pointer(cErr))
+		return libraryHandles{}, fmt.Errorf("failed to load shared library at %q: %s", libPath, C.GoString(cErr))
+	}
+
+	return (libraryHandles)(cLibHandles), nil
+}
+
+func (l *sharedLibraryLoader) Close(libHandle unsafe.Pointer) error {
+	var cErr *C.char
+
+	C.close_shared_library(libHandle, &cErr)
+	if cErr != nil {
+		defer C.free(unsafe.Pointer(cErr))
+		return fmt.Errorf("Close failed: %s", C.GoString(cErr))
+	}
+
+	return nil
+}
+
+func (l *sharedLibraryLoader) Run(runPtr *C.run_function_t, checkID string, initConfig string, instanceConfig string) error {
+	cID := C.CString(checkID)
+	defer C.free(unsafe.Pointer(cID))
+
+	cInitConfig := C.CString(initConfig)
+	defer C.free(unsafe.Pointer(cInitConfig))
+
+	cInstanceConfig := C.CString(instanceConfig)
+	defer C.free(unsafe.Pointer(cInstanceConfig))
+
+	var cErr *C.char
+
+	C.run_shared_library(runPtr, cID, cInitConfig, cInstanceConfig, l.aggregator, &cErr)
+	if cErr != nil {
+		defer C.free(unsafe.Pointer(cErr))
+		return fmt.Errorf("Run failed: %s", C.GoString(cErr))
+	}
+
+	return nil
+}
+
+func (l *sharedLibraryLoader) Version(versionPtr *C.version_function_t) (string, error) {
+	var cErr *C.char
+
+	cLibVersion := C.get_version_shared_library(versionPtr, &cErr)
+	if cErr != nil {
+		defer C.free(unsafe.Pointer(cErr))
+		return "", fmt.Errorf("Failed to get version: %s", C.GoString(cErr))
+	}
+
+	return C.GoString(cLibVersion), nil
+
+}
+
+func newSharedLibraryLoader(folderPath string) *sharedLibraryLoader {
+	return &sharedLibraryLoader{
+		folderPath: folderPath,
+		aggregator: C.get_aggregator(),
+	}
+}
+
+// mock of sharedLibraryLoader
+type mockSharedLibraryLoader struct{}
+
+func (ml *mockSharedLibraryLoader) Load(_ string) (libraryHandles, error) {
+	return libraryHandles{}, nil
+}
+
+func (ml *mockSharedLibraryLoader) Close(_ unsafe.Pointer) error {
+	return nil
+}
+
+func (ml *mockSharedLibraryLoader) Run(_ *C.run_function_t, _ string, _ string, _ string) error {
+	return nil
+}
+
+func (ml *mockSharedLibraryLoader) Version(_ *C.version_function_t) (string, error) {
+	return "mock_version", nil
+}
```

Comment:
The mock should be defined a separate file with `test` build tag

Depending on how it's used, you can use `mockery` to autogenerate an implementation for a given interface, which allows asserting expectation on calls
Current implementation is more a no-op loader than a mock really